In [1]:
library(Seurat)
library(MAST)

Attaching SeuratObject

Attaching sp

Loading required package: SingleCellExperiment

Loading required package: SummarizedExperiment

Loading required package: MatrixGenerics

Loading required package: matrixStats


Attaching package: ‘MatrixGenerics’


The following objects are masked from ‘package:matrixStats’:

    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
    colWeightedMeans, colWeightedMedians, colWeightedSds,
    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
    rowMadDiffs, rowMads, rowMaxs

In [2]:
library(dplyr)


Attaching package: ‘dplyr’


The following object is masked from ‘package:Biobase’:

    combine


The following objects are masked from ‘package:GenomicRanges’:

    intersect, setdiff, union


The following object is masked from ‘package:GenomeInfoDb’:

    intersect


The following objects are masked from ‘package:IRanges’:

    collapse, desc, intersect, setdiff, slice, union


The following objects are masked from ‘package:S4Vectors’:

    first, intersect, rename, setdiff, setequal, union


The following objects are masked from ‘package:BiocGenerics’:

    combine, intersect, setdiff, union


The following object is masked from ‘package:matrixStats’:

    count


The following objects are masked from ‘package:stats’:

    filter, lag


The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union




In [3]:
data_root <- '/lustre/scratch126/cellgen/team205/nh3/skin'
proj_root <- '~/FCA_Fetal_Skin_priv'

In [4]:
pooled_ad = readRDS(file.path(proj_root, "data", "pooled_fetal_organoid_adult_processed.rds"))

In [5]:
colnames(pooled_ad@meta.data)

[1] "sanger_id"                               
 [2] "chemistry"                               
 [3] "batch"                                   
 [4] "gender"                                  
 [5] "pcw"                                     
 [6] "sorting"                                 
 [7] "sample"                                  
 [8] "chemistry_sorting"                       
 [9] "cell_caller"                             
[10] "scrublet_score"                          
[11] "cluster_scrublet_score"                  
[12] "doublet_pval"                            
[13] "bh_doublet_pval"                         
[14] "nCounts_RNA"                             
[15] "nFeaturess_RNA"                          
[16] "percent_mito"                            
[17] "percent_ribo"                            
[18] "percent_hb"                              
[19] "independent_annotation_refined"          
[20] "independent_annotation_broad1"           
[21] "independent_annotation_broad2"           
[22] "independent_annotation_broad3"           
[23] "independent_annotation_broad4"           
[24] "independent_annotation_broad5"           
[25] "joint_annotation"                        
[26] "is_maternal"                             
[27] "fig1b_annotation"                        
[28] "independent_annotation_refined_20220201a"
[29] "independent_annotation_refined_20220201b"
[30] "independent_annotation_broad1_20220201"  
[31] "independent_annotation_broad2_20220201"  
[32] "independent_annotation_broad3_20220201"  
[33] "joint_annotation_20220201a"              
[34] "joint_annotation_20220201b"              
[35] "fig1b_annotation_20220201"               
[36] "joint_annotation_20220202"               
[37] "independent_annotation_refined_20220202" 
[38] "annot"                                   
[39] "day"                                     
[40] "sample_id"                               
[41] "version"                                 
[42] "week"                                    
[43] "nh3_annot1"                              
[44] "log1p_nCounts_RNA"                       
[45] "log1p_nFeaturess_RNA"                    
[46] "nCounts_RNA_mito"                        
[47] "nCounts_RNA_ribo"                        
[48] "nCounts_RNA_hb"                          
[49] "percent_top50"                           
[50] "pass_auto_filter"                        
[51] "scrublet_score_z"                        
[52] "bh_pval"                                 
[53] "scrublet_done"                           
[54] "doublet"                                 
[55] "seurat_clusters"                         
[56] "fsk_annot1_min30"                        
[57] "fsk_annot1_min30_prob"                   
[58] "fsk_annot2_min30"                        
[59] "fsk_annot2_min30_prob"                   
[60] "day8to13"                                
[61] "org_annot0"                              
[62] "org_annot0_prob"                         
[63] "org_annot30"                             
[64] "org_annot30_prob"                        
[65] "fsk_annot0"                              
[66] "fsk_annot0_prob"                         
[67] "fsk_annot30"                             
[68] "fsk_annot30_prob"                        
[69] "mad_prd"                                 
[70] "Status"                                  
[71] "Site"                                    
[72] "Site_old"                                
[73] "Tissue"                                  
[74] "Enrichment"                              
[75] "Location"                                
[76] "Sex"                                     
[77] "Age"                                     
[78] "stage"                                   
[79] "anno_final"                              
[80] "final"                                   
[81] "dataset"                                 
[82] "log2pcw"                                 
[83] "annot_early_late"                        
[84] "annot_early

In [35]:
pooled_ad@meta.data$annot %>% table()

.
             adult_Fibroblast                     adult_ILC 
                        17456                          7038 
           adult_Keratinocyte         adult_Langerhans cell 
                        53844                         12545 
  adult_Lymphatic endothelium              adult_Macrophage 
                         4924                          7205 
              adult_Mast cell              adult_Melanocyte 
                          557                          3682 
               adult_Monocyte              adult_Mural cell 
                         6685                          5006 
            adult_Plasma cell            adult_Schwann cell 
                           71                           269 
                 adult_T cell    adult_Vascular endothelium 
                        34774                         25785 
                    adult_cDC               fetal_Adipocyte 
                        15898                          3203 
                 fetal

In [6]:
df = data.frame(pooled_ad@meta.data[grepl("fetal|organoid", rownames(pooled_ad@meta.data)),])

In [7]:
unique(sapply(strsplit(rownames(df), split="-"), tail, n=1))

[1] "fetal"    "organoid"

In [19]:
df

sanger_id       chemistry
AAACCTGGTCAGTGGA-1-4834STDY7002879-fetal         4834STDY7002879 SC3Pv2   
AAAGATGGTCGATTGT-1-4834STDY7002879-fetal         4834STDY7002879 SC3Pv2   
AAAGCAAAGATGTGGC-1-4834STDY7002879-fetal         4834STDY7002879 SC3Pv2   
AAAGTAGCAGATCGGA-1-4834STDY7002879-fetal         4834STDY7002879 SC3Pv2   
AAAGTAGTCCGCATCT-1-4834STDY7002879-fetal         4834STDY7002879 SC3Pv2   
AAATGCCAGCTGCCCA-1-4834STDY7002879-fetal         4834STDY7002879 SC3Pv2   
AAATGCCTCGTGGTCG-1-4834STDY7002879-fetal         4834STDY7002879 SC3Pv2   
AACACGTAGAGAGCTC-1-4834STDY7002879-fetal         4834STDY7002879 SC3Pv2   
AACCATGAGATCTGCT-1-4834STDY7002879-fetal         4834STDY7002879 SC3Pv2   
AACCATGGTGCCTTGG-1-4834STDY7002879-fetal         4834STDY7002879 SC3Pv2   
AACTCAGCAGGACGTA-1-4834STDY7002879-fetal         4834STDY7002879 SC3Pv2   
AACTCAGCAGGTCTCG-1-4834STDY7002879-fetal         4834STDY7002879 SC3Pv2   
AACTCAGGTCAGCTAT-1-4834STDY7002879-fetal         4834STDY7002879 SC3Pv2   
AACTTTCAGTGCCATT-1-4834STDY7002879-fetal         4834STDY7002879 SC3Pv2   
AAGACCTCATGAAGTA-1-4834STDY7002879-fetal         4834STDY7002879 SC3Pv2   
AAGCCGCAGTGTCTCA-1-4834STDY7002879-fetal         4834STDY7002879 SC3Pv2   
AAGTCTGGTCGACTAT-1-4834STDY7002879-fetal         4834STDY7002879 SC3Pv2   
AAGTCTGTCGCCTGTT-1-4834STDY7002879-fetal         4834STDY7002879 SC3Pv2   
AATCCAGGTGTTTGTG-1-4834STDY7002879-fetal         4834STDY7002879 SC3Pv2   
AATCGGTAGTACGTAA-1-4834STDY7002879-fetal         4834STDY7002879 SC3Pv2   
AATCGGTAGTGGACGT-1-4834STDY7002879-fetal         4834STDY7002879 SC3Pv2   
AATCGGTTCCTAGAAC-1-4834STDY7002879-fetal         4834STDY7002879 SC3Pv2   
ACACCAATCCTCTAGC-1-4834STDY7002879-fetal         4834STDY7002879 SC3Pv2   
ACACCCTGTCGTTGTA-1-4834STDY7002879-fetal         4834STDY7002879 SC3Pv2   
ACACTGAAGCCAGAAC-1-4834STDY7002879-fetal         4834STDY7002879 SC3Pv2   
ACAGCTAGTAAACGCG-1-4834STDY7002879-fetal         4834STDY7002879 SC3Pv2   
ACATACGAGTGATCGG-1-4834STDY7002879-fetal         4834STDY7002879 SC3Pv2   
ACATCAGGTAAATACG-1-4834STDY7002879-fetal         4834STDY7002879 SC3Pv2   
ACATGGTCATCGGGTC-1-4834STDY7002879-fetal         4834STDY7002879 SC3Pv2   
ACCAGTAAGGCGCTCT-1-4834STDY7002879-fetal         4834STDY7002879 SC3Pv2   
⋮                                                ⋮               ⋮        
day_13_IEOdev_DSP_d13:TTTGGTTTCCAATGCAx-organoid nan             nan      
day_13_IEOdev_DSP_d13:TTTGATCTCTCTCGACx-organoid nan             nan      
day_13_IEOdev_DSP_d13:TTTGGTTGTCCAGTTAx-organoid nan             nan      
day_13_IEOdev_DSP_d13:TTTCGATTCATGCTAGx-organoid nan             nan      
day_13_IEOdev_DSP_d13:TTTGGTTTCGGTCGGTx-organoid nan             nan      
day_13_IEOdev_DSP_d13:TTTCCTCTCATAGACCx-organoid nan             nan      
day_13_IEOdev_DSP_d13:TTTGATCGTTAGGGTGx-organoid nan             nan      
day_13_IEOdev_DSP_d13:TTTGGTTGTGAATGATx-organoid nan             nan      
day_13_IEOdev_DSP_d13:TTTGATCGTAGAATGTx-organoid nan             nan      
day_13_IEOdev_DSP_d13:TTTGACTCATGACTCAx-organoid nan             nan      
day_13_IEOdev_DSP_d13:TTTGGTTTCTCCCAACx-organoid nan             nan      
day_13_IEOdev_DSP_d13:TTTGGAGGTGGAAGTCx-organoid nan             nan      
day_13_IEOdev_DSP_d13:TTTGTTGAGTCGCCCAx-organoid nan             nan      
day_13_IEOdev_DSP_d13:TTTGATCTCAAACGAAx-organoid nan             nan      
day_13_IEOdev_DSP_d13:TTTCCTCCACTAGGTTx-organoid nan             nan      
day_13_IEOdev_DSP_d13:TTTGGAGGTTATTCCTx-organoid nan             nan      
day_13_IEOdev_DSP_d13:TTTGATCCATCATCTTx-organoid nan             nan      
day_13_IEOdev_DSP_d13:TTTGGAGTCGAGCACCx-organoid nan             nan      
day_13_IEOdev_DSP_d13:TTTGGAGGTACAGAATx-organoid nan             nan      
day_13_IEOdev_DSP_d13:TTTGTTGGTACAGAATx-organoid nan             nan      
day_13_IEOdev_DSP_d13:TTTCGATGTCGACGCTx-organoid nan             nan      
day_13_IEOdev_DSP_d13:TTTGTTGTCCGAGCTGx-organoid nan             nan      

In [8]:
check_column_string = function(df, string) {
  sapply(df, function(x) sum(grepl(string, x))) > 0
}

# get the column names which contain the string
cols_with_string = names(df)[check_column_string(df, "Tip cell")]

# subset the data frame with the column names
df = df[cols_with_string]

In [9]:
cols_with_string

[1] "independent_annotation_refined"          
 [2] "joint_annotation"                        
 [3] "independent_annotation_refined_20220201a"
 [4] "independent_annotation_refined_20220201b"
 [5] "joint_annotation_20220201a"              
 [6] "joint_annotation_20220201b"              
 [7] "joint_annotation_20220202"               
 [8] "independent_annotation_refined_20220202" 
 [9] "fsk_annot1_min30"                        
[10] "fsk_annot2_min30"                        
[11] "org_annot0"                              
[12] "org_annot30"

In [11]:
for (c in colnames(df)){
    print(c)
    print(unique(df[[c]][grepl("Tip cell", df[[c]])]))
}

[1] "independent_annotation_refined"
[1] Tip cell (arterial)
79 Levels: ASDC Activated mac Adipocytes Angiogenic pericyte ... pDC
[1] "joint_annotation"
[1] Tip cell (arterial)
85 Levels: ASDC Activated mac Adipocytes Angiogenic pericyte ... pDC
[1] "independent_annotation_refined_20220201a"
[1] Tip cell (arterial)
79 Levels: ASDC Activated mac Adipocytes Arterial B cell ... pDC
[1] "independent_annotation_refined_20220201b"
[1] Tip cell (arterial)
78 Levels: ASDC Activated mac Adipocytes Arterial B cell ... pDC
[1] "joint_annotation_20220201a"
[1] Tip cell (arterial)
85 Levels: ASDC Activated mac Adipocytes Arterial B cell Basal ... pDC
[1] "joint_annotation_20220201b"
[1] Tip cell (arterial)
84 Levels: ASDC Activated mac Adipocytes Arterial B cell Basal ... pDC
[1] "joint_annotation_20220202"
[1] Tip cell (arterial)
84 Levels: ASDC Adipocytes Arterial B cell Basal CD4 T cell ... pDC
[1] "independent_annotation_refined_20220202"
[1] Tip cell (arterial)
78 Levels: ASDC Adipocytes Arter

In [16]:
length(df$org_annot30[grepl("Tip cell", df$org_annot30)])

[1] 130

In [56]:
unique(df$fsk_annot2_min30[grepl("Tip cell", df$fsk_annot2_min30)])

[1] Tip cell (arterial)
35 Levels: Adipocytes Angiogenic pericyte Basal Basal POSTN+ ... nan

<Font color="red">~130 cells are not enough to produce statistically robust differential expression results</Font>

In [115]:
pooled_ad$org_annot30[pooled_ad$fsk_annot2_min30 == "Tip cell (arterial)"]

AGGATCTAGTTGAAAC-1-2-organoid 
                    organoid_Tip cell (arterial) 
                   AGGGCTCCAAGCGCAA-1-2-organoid 
                    organoid_Tip cell (arterial) 
                   GTTACAGTCCGCTGTT-1-2-organoid 
                    organoid_Tip cell (arterial) 
                   TCGACCTGTACATACC-1-2-organoid 
                    organoid_Tip cell (arterial) 
                   CAATACGAGAAACTGT-1-4-organoid 
                    organoid_Tip cell (arterial) 
                   GTTAGACCAGTAGAGC-1-4-organoid 
                    organoid_Tip cell (arterial) 
                   GTCGTTCCAATGGGTG-1-5-organoid 
                    organoid_Tip cell (arterial) 
                   GTGAGGATCTAGTGTG-1-5-organoid 
                    organoid_Tip cell (arterial) 
                   GTTACCCCATCGTGGC-1-5-organoid 
                    organoid_Tip cell (arterial) 
                   TACCGAACAAAGGGTC-1-5-organoid 
                    organoid_Tip cell (arterial) 
                   TCAAGACCAACCAATC-1-5-organoid 
                    organoid_Tip cell (arterial) 
                   TTGACCCTCCCTTCCC-1-5-organoid 
                    organoid_Tip cell (arterial) 
                   TGAGCATCAGCAGTTT-1-6-organoid 
                    organoid_Tip cell (arterial) 
                   CTAATGGGTCTCAACA-1-7-organoid 
                    organoid_Tip cell (arterial) 
                   GCGAGAATCTGTTGAG-1-7-organoid 
                    organoid_Tip cell (arterial) 
                   GTATCTTCATGTTCCC-1-7-organoid 
                    organoid_Tip cell (arterial) 
                   TTCTTAGGTGGGTCAA-1-7-organoid 
                    organoid_Tip cell (arterial) 
                   AAACCCATCTCGACGG-1-8-organoid 
                    organoid_Tip cell (arterial) 
                   AAGCATCTCCCACAGG-1-8-organoid 
                    organoid_Tip cell (arterial) 
                   AAGCGTTGTTCTCCCA-1-8-organoid 
                    organoid_Tip cell (arterial) 
                   ACTTAGGGTTGTTGTG-1-8-organoid 
                    organoid_Tip cell (arterial) 
                   AGAGCAGTCGAGTGGA-1-8-organoid 
                    organoid_Tip cell (arterial) 
                   AGCTACAAGACAACTA-1-8-organoid 
                    organoid_Tip cell (arterial) 
                   ATCGTGATCGGTAGAG-1-8-organoid 
                    organoid_Tip cell (arterial) 
                   ATTACTCCACAGCTTA-1-8-organoid 
                    organoid_Tip cell (arterial) 
                   CAACGGCTCTCACTCG-1-8-organoid 
                    organoid_Tip cell (arterial) 
                   CATGCCTGTTCTAACG-1-8-organoid 
                    organoid_Tip cell (arterial) 
                   CATGCGGCAATGAAAC-1-8-organoid 
                    organoid_Tip cell (arterial) 
                   CCAATTTTCCGATGTA-1-8-organoid 
                    organoid_Tip cell (arterial) 
                   CCGAACGAGATTGACA-1-8-organoid 
                    organoid_Tip cell (arterial) 
                   CGGCAGTAGAGTGACC-1-8-organoid 
                    organoid_Tip cell (arterial) 
                   CGGGACTCACTGGATT-1-8-organoid 
                    organoid_Tip cell (arterial) 
                   CGTAAGTAGTAATCCC-1-8-organoid 
                    organoid_Tip cell (arterial) 
                   CTACCCATCCGTGTCT-1-8-organoid 
                    organoid_Tip cell (arterial) 
                   CTACCTGAGCTGTGCC-1-8-organoid 
                    organoid_Tip cell (arterial) 
                   CTGCCTACAGTGACCC-1-8-organoid 
                    organoid_Tip cell (arterial) 
                   CTGTGGGTCACTGTTT-1-8-organoid 
                    organoid_Tip cell (arterial) 
                   CTTCGGTCAGCGACAA-1-8-organoid 
                    organoid_Tip cell (arterial) 
                   GAGTCTATCTTGCGCT-1-8-organoid 
                    organoid_Tip cell (arterial) 
                   GGAATCTCATAGCACT-1-8-organoid 
                    organoid_Tip cell (arterial) 
                   

In [59]:
pooled_ad$capillary_arterioles = NA
pooled_ad$capillary_arterioles[pooled_ad$fsk_annot2_min30 == "Tip cell (arterial)"] = "fetal_capillary_arterioles"
pooled_ad$capillary_arterioles[pooled_ad$org_annot30 == "organoid_Tip cell (arterial)"] = "organoid_capillary_arterioles"

ERROR: Error in `[[<-`(`*tmp*`, pooled_ad$fsk_annot2_min30 == "Tip cell (arterial)", : attempt to select more than one element in vectorIndex


In [54]:
pooled_ad$capillary_arterioles[!is.na(pooled_ad$capillary_arterioles)]

AGGATCTAGTTGAAAC-1-2-organoid 
                 "organoid_capillary_arterioles" 
                   AGGGCTCCAAGCGCAA-1-2-organoid 
                 "organoid_capillary_arterioles" 
                   GTTACAGTCCGCTGTT-1-2-organoid 
                 "organoid_capillary_arterioles" 
                   TCGACCTGTACATACC-1-2-organoid 
                 "organoid_capillary_arterioles" 
                   CAATACGAGAAACTGT-1-4-organoid 
                 "organoid_capillary_arterioles" 
                   GTTAGACCAGTAGAGC-1-4-organoid 
                 "organoid_capillary_arterioles" 
                   GTCGTTCCAATGGGTG-1-5-organoid 
                 "organoid_capillary_arterioles" 
                   GTGAGGATCTAGTGTG-1-5-organoid 
                 "organoid_capillary_arterioles" 
                   GTTACCCCATCGTGGC-1-5-organoid 
                 "organoid_capillary_arterioles" 
                   TACCGAACAAAGGGTC-1-5-organoid 
                 "organoid_capillary_arterioles" 
                   TCAAGACCAACCAATC-1-5-organoid 
                 "organoid_capillary_arterioles" 
                   TTGACCCTCCCTTCCC-1-5-organoid 
                 "organoid_capillary_arterioles" 
                   TGAGCATCAGCAGTTT-1-6-organoid 
                 "organoid_capillary_arterioles" 
                   CTAATGGGTCTCAACA-1-7-organoid 
                 "organoid_capillary_arterioles" 
                   GCGAGAATCTGTTGAG-1-7-organoid 
                 "organoid_capillary_arterioles" 
                   GTATCTTCATGTTCCC-1-7-organoid 
                 "organoid_capillary_arterioles" 
                   TGTTCCGAGAGGACGG-1-7-organoid 
                 "organoid_capillary_arterioles" 
                   TTCTTAGGTGGGTCAA-1-7-organoid 
                 "organoid_capillary_arterioles" 
                   AAACCCATCTCGACGG-1-8-organoid 
                 "organoid_capillary_arterioles" 
                   AAGCATCTCCCACAGG-1-8-organoid 
                 "organoid_capillary_arterioles" 
                   AAGCGTTGTTCTCCCA-1-8-organoid 
                 "organoid_capillary_arterioles" 
                   ACTTAGGGTTGTTGTG-1-8-organoid 
                 "organoid_capillary_arterioles" 
                   AGAGCAGTCGAGTGGA-1-8-organoid 
                 "organoid_capillary_arterioles" 
                   AGCTACAAGACAACTA-1-8-organoid 
                 "organoid_capillary_arterioles" 
                   ATCGTGATCGGTAGAG-1-8-organoid 
                 "organoid_capillary_arterioles" 
                   ATTACTCCACAGCTTA-1-8-organoid 
                 "organoid_capillary_arterioles" 
                   CAACGGCTCTCACTCG-1-8-organoid 
                 "organoid_capillary_arterioles" 
                   CATGCCTGTTCTAACG-1-8-organoid 
                 "organoid_capillary_arterioles" 
                   CATGCGGCAATGAAAC-1-8-organoid 
                 "organoid_capillary_arterioles" 
                   CCAATTTTCCGATGTA-1-8-organoid 
                 "organoid_capillary_arterioles" 
                   CCGAACGAGATTGACA-1-8-organoid 
                 "organoid_capillary_arterioles" 
                   CGGCAGTAGAGTGACC-1-8-organoid 
                 "organoid_capillary_arterioles" 
                   CGGGACTCACTGGATT-1-8-organoid 
                 "organoid_capillary_arterioles" 
                   CGTAAGTAGTAATCCC-1-8-organoid 
                 "organoid_capillary_arterioles" 
                   CTACCCATCCGTGTCT-1-8-organoid 
                 "organoid_capillary_arterioles" 
                   CTACCTGAGCTGTGCC-1-8-organoid 
                 "organoid_capillary_arterioles" 
                   CTGCCTACAGTGACCC-1-8-organoid 
                 "organoid_capillary_arterioles" 
                   CTGTGGGTCACTGTTT-1-8-organoid 
                 "organoid_capillary_arterioles" 
                   CTTCGGTCAGCGACAA-1-8-organoid 
                 "organoid_capillary_arterioles" 
                   GAGTCTATCTTGCGCT-1-8-organoid 
                 "organoid_capillary_arterioles" 
                   

In [47]:
# Subset on the `$capillary_arterioles` value in the object meta data
# cells_1 = subset(pooled_ad, subset = pooled_ad$capillary_arterioles == "fetal_capillary_arterioles")
cells_1 = colnames(pooled_ad[, pooled_ad$capillary_arterioles == "fetal_capillary_arterioles"])
# cells_2 = subset(pooled_ad, subset = pooled_ad$capillary_arterioles == "organoid_capillary_arterioles")
cells_2 = colnames(pooled_ad[, pooled_ad$capillary_arterioles == "organoid_capillary_arterioles"])

In [48]:
cells_1

[1] "TCTCACGAGACTTGTC-1-9-organoid"  "AGGGTGAGTTCCATGA-1-11-organoid"
[3] "CAACTAGCATCGGTTA-1-11-organoid"

In [36]:
diff_exp_genes = MAST::MASTDETest(pooled_ad, cells_1, cells_2)

ERROR: Error: 'MASTDETest' is not an exported object from 'namespace:MAST'


In [65]:
?FindMarkers

FindMarkers               package:Seurat               R Documentation

_G_e_n_e _e_x_p_r_e_s_s_i_o_n _m_a_r_k_e_r_s _o_f _i_d_e_n_t_i_t_y _c_l_a_s_s_e_s

_D_e_s_c_r_i_p_t_i_o_n:

     Finds markers (differentially expressed genes) for identity
     classes

_U_s_a_g_e:

     FindMarkers(object, ...)
     
     ## Default S3 method:
     FindMarkers(
       object,
       slot = "data",
       counts = numeric(),
       cells.1 = NULL,
       cells.2 = NULL,
       features = NULL,
       logfc.threshold = 0.25,
       test.use = "wilcox",
       min.pct = 0.1,
       min.diff.pct = -Inf,
       verbose = TRUE,
       only.pos = FALSE,
       max.cells.per.ident = Inf,
       random.seed = 1,
       latent.vars = NULL,
       min.cells.feature = 3,
       min.cells.group = 3,
       pseudocount.use = 1,
       fc.results = NULL,
       ...
     )
     
     ## S3 method for class 'Assay'
     FindMarkers(
       object,
       slot = "data",
       cells.1 = NULL,
       cells.2 = NULL,
       features = NULL,
       logfc.threshold = 0.25,
       test.use = "wilcox",
       min.pct = 0.1,
       min.diff.pct = -Inf,
       verbose = TRUE,
       only.pos = FALSE,
       max.cells.per.ident = Inf,
       random.seed = 1,
       latent.vars = NULL,
       min.cells.feature = 3,
       min.cells.group = 3,
       pseudocount.use = 1,
       mean.fxn = NULL,
       fc.name = NULL,
       base = 2,
       ...
     )
     
     ## S3 method for class 'DimReduc'
     FindMarkers(
       object,
       cells.1 = NULL,
       cells.2 = NULL,
       features = NULL,
       logfc.threshold = 0.25,
       test.use = "wilcox",
       min.pct = 0.1,
       min.diff.pct = -Inf,
       verbose = TRUE,
       only.pos = FALSE,
       max.cells.per.ident = Inf,
       random.seed = 1,
       latent.vars = NULL,
       min.cells.feature = 3,
       min.cells.group = 3,
       pseudocount.use = 1,
       mean.fxn = rowMeans,
       ...
     )
     
     ## S3 method for class 'Seurat'
     FindMarkers(
       object,
       ident.1 = NULL,
       ident.2 = NULL,
       group.by = NULL,
       subset.ident = NULL,
       assay = NULL,
       slot = "data",
       reduction = NULL,
       features = NULL,
       logfc.threshold = 0.25,
       test.use = "wilcox",
       min.pct = 0.1,
       min.diff.pct = -Inf,
       verbose = TRUE,
       only.pos = FALSE,
       max.cells.per.ident = Inf,
       random.seed = 1,
       latent.vars = NULL,
       min.cells.feature = 3,
       min.cells.group = 3,
       pseudocount.use = 1,
       mean.fxn = NULL,
       fc.name = NULL,
       base = 2,
       ...
     )
     
_A_r_g_u_m_e_n_t_s:

  object: An object

     ...: Arguments passed to other methods and to specific DE methods

    slot: Slot to pull data from; note that if ‘test.use’ is
          "negbinom", "poisson", or "DESeq2", ‘slot’ will be set to
          "counts"

  counts: Count matrix if using scale.data for DE tests. This is used
          for computing pct.1 and pct.2 and for filtering features
          based on fraction expressing

 cells.1: Vector of cell names belonging to group 1

 cells.2: Vector of cell names belonging to group 2

features: Genes to test. Default is to use all genes

logfc.threshold: Limit testing to genes which show, on average, at
          least X-fold difference (log-scale) between the two groups of
          cells. Default is 0.25 Increasing logfc.threshold speeds up
          the function, but can miss weaker signals.

test.use: Denotes which test to use. Available options are:

            • "wilcox" : Identifies differentially expressed genes
              between two groups of cells using a Wilcoxon Rank Sum
              test (default)

            • "bimod" : Likelihood-ratio test for single cell gene
              expression, (McDavid et al., Bioinformatics, 2013)

            • "roc" : Identifies 'markers' of gene expression using ROC
   

In [73]:
# Perform differential expression analysis
de_results <- FindMarkers(pooled_ad, cells_1, cells_2, test.use = "MAST", logfc.threshold = 0.25)


Done!

Combining coefficients and standard errors

Calculating log-fold changes

Calculating likelihood ratio tests

Refitting on reduced model...


Done!



In [84]:
de_results[order(abs(de_results[, "avg_log2FC"]), decreasing = TRUE), ]


p_val        avg_log2FC pct.1 pct.2 p_val_adj
PTN      0.0135631530  3.481791  0.333 0.169 1        
UBE2C    0.0037808086  3.038616  0.667 0.108 1        
COL3A1   0.0002788539  2.802564  0.333 0.431 1        
PTTG1    0.0157257432  2.527251  0.667 0.138 1        
ARL6IP1  0.0008037031  2.104016  0.667 0.654 1        
SFRP2    0.0825681795  2.083583  0.333 0.092 1        
RPS4Y1   0.0341172639 -2.015701  0.000 0.592 1        
MGP      0.1816760752  2.002652  0.333 0.115 1        
CDK1     0.0265915497  1.967298  0.667 0.123 1        
HMGB2    0.0073991748  1.916832  1.000 0.400 1        
LDHA     0.0558685935 -1.875169  0.333 0.900 1        
UBE2S    0.0000481710  1.874852  0.667 0.400 1        
EIF4EBP1 0.0205379982 -1.773034  0.333 0.923 1        
BCAP31   0.0027296316 -1.759390  0.000 0.823 1        
ACKR3    0.0072208498 -1.758801  0.000 0.754 1        
PCLAF    0.0132228222  1.744362  1.000 0.215 1        
HMGN2    0.0084806258  1.742515  1.000 0.808 1        
MKI67    0.0052457153  1.730225  0.667 0.108 1        
PRRX1    0.0819901800  1.725540  0.333 0.077 1        
PRSS2    0.1368676593 -1.720786  1.000 0.746 1        
SMAD1    0.0024005704 -1.701940  0.000 0.831 1        
CDC20    0.0135808127  1.701473  0.667 0.038 1        
CKS2     0.0054080547  1.697888  0.667 0.438 1        
P4HB     0.0210497524 -1.686653  0.333 0.915 1        
CENPN    0.0031935513  1.686444  0.667 0.154 1        
CCNA2    0.0216060250  1.665068  0.667 0.069 1        
CRABP1   0.0029080452  1.655757  0.333 0.292 1        
ESM1     0.2845798816 -1.628091  0.333 0.738 1        
CENPW    0.0078197942  1.628023  0.667 0.277 1        
C2orf69  0.0048087666  1.612651  0.667 0.146 1        
⋮        ⋮            ⋮          ⋮     ⋮     ⋮        
SARAF    0.8394885    -0.2511795 0.667 0.815 1        
IDS      0.7640309     0.2511068 0.667 0.477 1        
ARRB2    0.5680651     0.2511033 0.333 0.100 1        
RIOK1    0.3045630    -0.2510549 0.000 0.223 1        
SRD5A3   0.9271491    -0.2510266 0.333 0.415 1        
ELMOD2   0.2932709    -0.2510231 0.000 0.231 1        
ARMC7    0.6083595     0.2510068 0.333 0.115 1        
HAUS2    0.3283637    -0.2509677 0.000 0.208 1        
FAM136A  0.8431908    -0.2509642 0.333 0.500 1        
BAK1     0.2155462    -0.2509196 0.000 0.292 1        
E2F1     0.3338776     0.2508956 0.333 0.054 1        
CCDC85C  0.2422634    -0.2508947 0.000 0.269 1        
PSAT1    0.5969442     0.2508419 0.333 0.108 1        
KDELC2   0.6733051     0.2508333 0.333 0.169 1        
KAT6B    0.3045630    -0.2507993 0.000 0.223 1        
VKORC1L1 0.2155462    -0.2507013 0.000 0.292 1        
CD58     0.6240221     0.2506750 0.333 0.123 1        
ME3      0.3409089    -0.2506518 0.000 0.200 1        
RANBP6   0.2616326    -0.2506493 0.000 0.254 1        
TAPBPL   0.6733858     0.2506337 0.333 0.154 1        
MSMP     0.3813766    -0.2505910 0.000 0.177 1        
SLC15A4  0.6375639     0.2505787 0.333 0.131 1        
NUS1     0.2517845    -0.2505175 0.000 0.262 1        
TERF2    0.2330574    -0.2504391 0.000 0.277 1        
ARL15    0.3045630    -0.2503660 0.000 0.223 1        
IPO7     0.8205137    -0.2503627 0.333 0.515 1        
WDR6     0.5869650     0.2503068 0.333 0.346 1        
MUL1     0.2330574    -0.2502621 0.000 0.277 1        
OPA3     0.2932709    -0.2501218 0.000 0.231 1        
MTMR2    0.4871587     0.2501110 0.333 0.385 1

In [78]:
# Extract the differentially expressed genes
sig_genes <- rownames(de_results[de_results$p_val < 0.05, ])

In [82]:
sig_genes

[1] "UBE2S"      "KIF18B"     "COL3A1"     "SCUBE2"     "UBE2H"     
  [6] "ARL6IP1"    "ZNF706"     "STX3"       "CHID1"      "DTYMK"     
 [11] "PSMB6"      "SMAD1"      "GABPB1"     "PPP4R2"     "BCAP31"    
 [16] "CRABP1"     "COL1A2"     "CENPN"      "WRB"        "UBE2C"     
 [21] "RALB"       "CDCA3"      "C2orf69"    "TEN1"       "MKI67"     
 [26] "CKS2"       "ERGIC3"     "HIGD2A"     "SERPINB6"   "ACKR3"     
 [31] "PPP1CB"     "KIF2C"      "HMGB2"      "ZSCAN21"    "IL17RC"    
 [36] "GLCCI1"     "VPS37D"     "YIF1B"      "CENPW"      "PFDN4"     
 [41] "DBF4"       "CUL3"       "CDCA8"      "HMGN2"      "PXMP4"     
 [46] "NONO"       "PNRC1"      "STAT1"      "TROAP"      "BTF3"      
 [51] "GTSE1"      "NDC80"      "MED29"      "MIGA1"      "OSBPL9"    
 [56] "ZNF558"     "DDX3X"      "SBDS"       "SLC4A7"     "SYNCRIP"   
 [61] "C11orf74"   "PRDX1"      "CD163L1"    "KIF23"      "SPPL3"     
 [66] "CCDC167"    "MAP4"       "RAPGEF5"    "PPP2R2D"    "TYMS"      
 [71] "KNL1"       "RASGRP3"    "PCLAF"      "PTN"        "CDC20"     
 [76] "TAZ"        "AC010240.3" "AC019171.1" "AC133550.2" "AC146944.4"
 [81] "AC245123.1" "AF111167.2" "AL096865.1" "AL121821.1" "ANKRD29"   
 [86] "AP000866.1" "C1QTNF7"    "CAPN6"      "CCDC136"    "CCDC177"   
 [91] "CELF4"      "CXorf67"    "DTX3"       "DYTN"       "ENPP1"     
 [96] "EYA4"       "HMGCLL1"    "LINC00626"  "LINC01173"  "LTC4S"     
[101] "MRGPRF"     "NPHS1"      "PAPPA2"     "PLSCR4"     "PNMA6A"    
[106] "PRR19"      "SRCIN1"     "STAM-AS1"   "SUSD4"      "SYT14"     
[111] "TMEM59L"    "VCAM1"      "ZFP3"       "ZIC4"       "AC008496.2"
[116] "AC068790.9" "AC245052.4" "AL121718.1" "AL359715.2" "AP001363.2"
[121] "HIST1H3B"   "TRPM1"      "Z99774.1"   "AC090627.1" "GALNT9"    
[126] "IL10RB-DT"  "KCNE5"      "MIAT"       "NDUFA13"    "FOXM1"     
[131] "TNRC6B"     "PFDN1"      "RASSF2"     "RBM33"      "KCNQ1OT1"  
[136] "PTTG1"      "SKA3"       "PRC1"       "ASXL1"      "AKAP13"    
[141] "LIPA"       "TECR"       "NCAPG"      "TFPT"       "AMD1"      
[146] "QPCT"       "DIAPH3"     "EIF4A2"     "ISG15"      "NAMPT"     
[151] "CCDC25"     "NAT14"      "RPA1"       "CNTROB"     "DLEU2"     
[156] "KDM4A"      "OSER1"      "ANLN"       "BGN"        "ARHGEF28"  
[161] "EIF4EBP1"   "SAE1"       "SNRPE"      "P4HB"       "POMT2"     
[166] "APLP2"      "CCNA2"      "TMEM60"     "ADIPOR1"    "LSM8"      
[171] "PSMB9"      "TNFRSF1A"   "HNRNPR"     "ILK"        "SPHK1"     
[176] "CMPK1"      "PFKL"       "RNPS1"      "CCNB1"      "PRNP"      
[181] "EIF2S2"     "PPM1F"      "KIAA1143"   "KIF14"      "SOGA1"     
[186] "MAPKAPK5"   "DDX19B"     "LAMB2"      "TUBB"       "CENPF"     
[191] "RHEB"       "C12orf45"   "RABGAP1"    "CDK1"       "FNBP1L"    
[196] "LYRM4"      "U2SURP"     "PLK1"       "TOP2A"      "FBRS"      
[201] "STAU2"      "SMARCB1"    "RAP1A"      "BAALC"      "DLL4"      
[206] "MTSS1"      "RIT1"       "ROCK2"      "ANAPC5"     "MXRA7"     
[211] "TMEM106C"   "CHMP3"      "G3BP2"      "GOLIM4"     "SEC23A"    
[216] "POLR2L"     "CPE"        "GPR161"     "AURKB"      "FREM1"     
[221] "FAM50A"     "STARD4"     "AP001505.1" "ARHGEF4"    "B3GALT1"   
[226] "BNC2"       "CD74"       "CPED1"      "GRID2"      "RND2"      
[231] "THRB"       "TNMD"       "AIF1"       "AL163051.1" "AL354836.1"
[236] "AL590560.2" "ASXL3"      "BARX1"      "BMP5"       "C19orf38"  
[241] "CDON"       "EPHA3"      "EPS8"       "FBXO2"      "FGF13"     
[246] "FSTL5"      "GAS7"       "LRRC4B"     "NHS"        "NKD1"      
[251] "PALM2"      "PHACTR3"    "PLPPR3"     "SAMD11"     "SIX1"      
[256] "SULF1"      "TCEAL5"     "ZNF700"     "AC018695.2" "AC022306.2"
[261] "BDNF-AS"    "BICC1"      "CDH23"      "COL13A1"    "GDF10"     
[266] "GXYLT2"     "JAKMIP2"    "KIF18A"     "NRK"        "STXBP2"    
[271] "Z97832.2"   "HCG17"      "AC110285.2" "AL731661.1" "ARHGEF39"  
[276] "C17orf53"   "C3orf67"    "C8orf88"    "CD274"      "E2F8"      
[281] "ERC2"       "HLA-DM

In [69]:
?assay

SummarizedExperiment-classpackage:SummarizedExperiment R Documentation

_S_u_m_m_a_r_i_z_e_d_E_x_p_e_r_i_m_e_n_t _o_b_j_e_c_t_s

_D_e_s_c_r_i_p_t_i_o_n:

     The SummarizedExperiment class is a matrix-like container where
     rows represent features of interest (e.g. genes, transcripts,
     exons, etc...)  and columns represent samples (with sample data
     summarized as a DataFrame). A SummarizedExperiment object contains
     one or more assays, each represented by a matrix-like object of
     numeric or other mode.

     Note that SummarizedExperiment is the parent of the
     RangedSummarizedExperiment class which means that all the methods
     documented below also work on a RangedSummarizedExperiment object.

_U_s_a_g_e:

     ## Constructor
     
     # See ?RangedSummarizedExperiment for the constructor function.
     
     ## Accessors
     
     assayNames(x, ...)
     assayNames(x, ...) <- value
     assays(x, withDimnames=TRUE, ...)
     assays(x, withDimnames=TRUE, ...) <- value
     assay(x, i, withDimnames=TRUE, ...)
     assay(x, i, withDimnames=TRUE, ...) <- value
     rowData(x, use.names=TRUE, ...)
     rowData(x, ...) <- value
     colData(x, ...)
     colData(x, ...) <- value
     #dim(x)
     #dimnames(x)
     #dimnames(x) <- value
     
     ## Quick colData access
     
     ## S4 method for signature 'SummarizedExperiment'
     x$name
     ## S4 replacement method for signature 'SummarizedExperiment'
     x$name <- value
     ## S4 method for signature 'SummarizedExperiment,ANY,missing'
     x[[i, j, ...]]
     ## S4 replacement method for signature 'SummarizedExperiment,ANY,missing'
     x[[i, j, ...]] <- value
     
     ## Subsetting
     
     ## S4 method for signature 'SummarizedExperiment'
     x[i, j, ..., drop=TRUE]
     ## S4 replacement method for signature 'SummarizedExperiment,ANY,ANY,SummarizedExperiment'
     x[i, j] <- value
     ## S4 method for signature 'SummarizedExperiment'
     subset(x, subset, select, ...)
     
     ## Combining
     
     ## S4 method for signature 'SummarizedExperiment'
     cbind(..., deparse.level=1)
     ## S4 method for signature 'SummarizedExperiment'
     rbind(..., deparse.level=1)
     
     ## On-disk realization
     ## S4 method for signature 'SummarizedExperiment'
     realize(x, BACKEND=getAutoRealizationBackend())
     
_A_r_g_u_m_e_n_t_s:

       x: A SummarizedExperiment object.

     ...: For ‘assay’, arguments in ‘...’ are forwarded to ‘assays’.

          For ‘cbind’, ‘rbind’, ‘...’ contains SummarizedExperiment
          objects to be combined.

          For other accessors, ignored.

   value: An object of a class specified in the S4 method signature or
          as outlined in ‘Details’.

    i, j: For ‘assay’, ‘assay<-’, ‘i’ is an integer or numeric scalar;
          see ‘Details’ for additional constraints.

          For ‘[,SummarizedExperiment’, ‘[,SummarizedExperiment<-’,
          ‘i’, ‘j’ are subscripts that can act to subset the rows and
          columns of ‘x’, that is the ‘matrix’ elements of ‘assays’.

          For ‘[[,SummarizedExperiment’, ‘[[<-,SummarizedExperiment’,
          ‘i’ is a scalar index (e.g., ‘character(1)’ or ‘integer(1)’)
          into a column of ‘colData’.

    name: A symbol representing the name of a column of ‘colData’.

withDimnames: A ‘logical(1)’, indicating whether the dimnames of the
          SummarizedExperiment object should be applied (i.e.  copied)
          to the extracted assays. More precisely, setting
          ‘withDimnames=FALSE’ in the _getter_ returns the assays
          _as-is_ whereas setting ‘withDimnames=FALSE’ return them with
          possibly modified dimnames.

          Setting ‘withDimnames=FALSE’ in the _setter_ (‘assays<-’) is
          required when the dimnames on the supplied assays are not
          identical to the dimnames on the SummarizedExperiment object;
          it does not influence actual assignmen

In [93]:
pooled_ad@assays

$RNA
Assay data with 46794 features for 506107 cells
First 10 features:
 A1BG, A1BG-AS1, A1CF, A2M, A2M-AS1, A2ML1, A2ML1-AS1, A2ML1-AS2,
A3GALT2, A4GALT 


In [107]:
cells_1 %>% head()

[1] "TCTCACGAGACTTGTC-1-9-organoid"  "AGGGTGAGTTCCATGA-1-11-organoid"
[3] "CAACTAGCATCGGTTA-1-11-organoid"

In [102]:
# Convert the Seurat object to a matrix for input to MAST
mat <- pooled_ad@assays[rownames(pooled_ad@assays) %in% c(cells_1, cells_2), ]

ERROR: Error in pooled_ad@assays[rownames(pooled_ad@assays) %in% c(cells_1, cells_2), : incorrect number of dimensions


In [87]:
# Specify the model for MAST
mod <- model.matrix(~0 + condition, data = colData(pooled_ad@assays))
colnames(mod) <- levels(colData(pooled_ad)$condition)

ERROR: Error in (function (classes, fdef, mtable) : unable to find an inherited method for function ‘colData’ for signature ‘"Seurat"’


In [ ]:
# Run MAST
fit <- mast(mat, mod)

# Results interpretation: Extract the differentially expressed genes
res <- tidy(fit)
res_sig <- res[res$qval < 0.05, ]

# Print the top 10 differentially expressed genes
head(res_sig[order(res_sig$logFC), ], 10)